### ISSUE TITLE
Automatic address allocation does not work for IPv6 [::]

### ISSUE THREAD

**Issue by oliv3r** [systemd/systemd#9955]
**systemd version the issue has been seen with**
> 231 - 238

**Used distribution**
Debian, Gentoo and CoreOS

**Expected behaviour you didn't see**
 > I get an fc00::/7 auto generated ULA IP address assigned

**Unexpected behaviour you saw**
 > Various errors and crashes (see below), generally not working.

**Steps to reproduce the problem**
As I tried this various times on my live gentoo server, I switched to using a VM as I couldn't get it to work. I tried it on my Debian system (buster, desktop), but blamed potentially network-manager interfering (which turns out was not the case).
I booted up an 'alpha' version of coreos in a VM to ensure I could get the latest possible systemd version, 238.

So I made various versions of /etc/systemd/network/30-eth0.network with the following results, here the first version:
`
[Match]
MACAddress=52:54:00:*

[Network]
LinkLocalAddressing=yes
Description="IPv6 test"
DHCP=yes
Address=[::]
Address=0.0.0.0
`
(I added the ipv4 address to see if it was working at all, for the below variations, I left it out)

The above config results in the following:
```
Aug 27 19:38:49 localhost systemd-networkd[915]: /etc/systemd/network/30-eth0.network:8: Address is invalid, ignoring assignment: [::]
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Flags change: +UP +LOWER_UP +RUNNING +MULTICAST +BROADCAST
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Link 2 added
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: udev initialized link
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Saved original MTU: 1500
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Adding address: fe80::5054:ff:fe76:4f86/64 (valid forever)
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Gained IPv6LL
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Adding address: 10.2.10.138/16 (valid for 6min 5s)
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Adding address: 10.0.0.1/8 (valid forever)
Aug 27 19:38:49 localhost systemd-networkd[915]: eth0: Adding address: 169.254.6.149/16 (valid forever)
```

As can be seen, [::] is being ignored, despite what's being written in the [manual](https://github.com/systemd/systemd/blob/master/man/systemd.network.xml#L484)

When looking at the [source](https://github.com/systemd/systemd/blob/master/src/network/networkd-address.c#L746) which causes the error, it appears that this is triggered from the configuration file parser. The documented [::] [parser](https://github.com/systemd/systemd/blob/master/src/network/networkd-address.c#L479) or at least notion off, only seems to be called from other functions which don't seem to related to the address parsing from file:
```
src/network/networkd-dhcp4.c:     r = address_configure(addr, link, dhcp4_address_handler, true);
src/network/networkd-dhcp6.c:     r = address_configure(addr, link, dhcp6_address_handler, true);
src/network/networkd-ipv4ll.c:       r = address_configure(ll_addr, link, ipv4ll_address_handler, false);
src/network/networkd-link.c:          r = address_configure(ad, link, address_handler, false);
src/network/networkd-ndisc.c:       r = address_configure(address, link, ndisc_netlink_handler, true);
```
so I'm inclined to think that's less relevant.

Going back to the configuration file parser, the failing function, in_addr_from_string_auto, is really just a wrapper around inet_pton, which is does state that :: gets [silently ignored](https://github.com/git/git/blob/master/compat/inet_pton.c#L97) (note that this is the git compat inet_pton, couldn't find a quick reference of the real one).

What I think happens is, that we tell the user to enter [::] which gets passed to inet_ptons which ignores the [ and thus returns invalid.

Switching to Address=::, which seems logical as inet_pton would accept that, yields not [unexpected](https://github.com/systemd/systemd/blob/master/src/network/networkd-manager.c#L1685) an error:
`
Assertion 'prefixlen > 0' failed at ../systemd-238/src/network/networkd-manager.c:1698, function manager_address_pool_acquire(). Aborting.
`

Of course, trying the next permutation, Address=::/7 'works' by satisfying both the prefixlen requirement and the inet_pton requirement, but not satisfactory as it sets the IPv6 address to fc00::1/7.

So either the documentation is wrong, in that ::/7 should be used, but then in turn, the generated IP address is wrong, as the fc00/7 subnet should generate a [ULA](https://en.wikipedia.org/wiki/Unique_local_address) IP, which more or less forces the IP into the fd00/8 subnet due to the required bit having to be set and also be randomly generated as per [RFC4193](https://tools.ietf.org/html/rfc4193) and thus ::/7 currently generates non-compliant address.

Or the implementation is wrong, in that [::] is passed straight to inet_pton, which is wrong as it needs the [ (and ]) stripped so that inet_pton happily approves the general IP and in turn, systemd can generate the required ULA IP using the random algorithm as described by [RFC4193](https://tools.ietf.org/html/rfc4193). But I haven't traced that bit out if systemd even those that in any case.

In either case, the manual does state, that [::] should [generate](https://github.com/systemd/systemd/blob/master/man/systemd.network.xml#L484) address and I cannot imagine it not being compliant.

P.S. I apologize beforehand for repeating and overstating things, it's quite late and I'm really tired ;)

**Comment by yuwata** [comment]
@oliv3r Fix is waiting in #11770. If possible, please test the PR. Thank you.

**Comment by oliv3r** [comment]
I will try it at some time, as I still want to debug #5972 hopefully before the next release :) but I doubt time is my friend here
